### PR TITLE
gh-126807: pygettext: Do not attempt to extract messages from function definitions.

### PR DESCRIPTION
**Pull request by tomasr8**
Fixes a bug where pygettext would attempt to extract a message from a code like this:

```python
def _(x): pass
```

This is because pygettext only looks at one token at a time and `_(x)` looks like a function call.

However, since `x` is not a string literal, it would erroneously issue a warning.

This PR fixes that by keeping track of the previous token and checking if it's `def` or `class`.

* Issue: gh-126807

### CODE REVIEW

File: Lib/test/test_tools/test_i18n.py [hunk 1]
```diff
@@ -87,17 +87,26 @@ def assert_POT_equal(self, expected, actual):
...
+            res = assert_python_ok('-Xutf8', self.script, *args, filename)
+            if strict:
+                self.assertEqual(res.err, b'')
             with open('messages.pot', encoding='utf-8') as fp:
                 data = fp.read()
         return self.get_msgids(data)
 
+    def extract_docstrings_from_str(self, module_content):
+        """Return all docstrings extracted from module_content."""
+        return self.extract_from_str(module_content, args=('--docstrings',), strict=False)
```
Comment (tomasr8) [inline Lib/test/test_tools/test_i18n.py:105]: Some docstring tests cause warnings, so I keep `strict=False` here.

File: Tools/i18n/pygettext.py [hunk 2]
```diff
@@ -1,11 +1,10 @@
 #! /usr/bin/env python3
-# -*- coding: iso-8859-1 -*-
```
Comment (tomasr8) [inline Tools/i18n/pygettext.py]: There's no other file that uses this encoding, I think it's safe (and more practical) to use utf-8.

File: Tools/i18n/pygettext.py [hunk 3]
```diff
@@ -341,7 +343,10 @@ def __waiting(self, ttype, tstring, lineno):
             if ttype == tokenize.NAME and tstring in ('class', 'def'):
                 self.__state = self.__suiteseen
                 return
-        if ttype == tokenize.NAME and tstring in opts.keywords:
+        if (
+            ttype == tokenize.NAME and tstring in opts.keywords
+            and (not self.__prev_token or not _is_def_or_class_keyword(self.__prev_token))
+        ):
```
Comment (tomasr8) [inline Tools/i18n/pygettext.py]: The new logic is, if we see one of the gettext keywords __and__ the previous token is not `def` or `class`, only then we transition to `__keywordseen`.

File: Tools/i18n/pygettext.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Tools/i18n/pygettext.py]: This is not related change, so please keep the coding cookie.

File: Tools/i18n/pygettext.py [hunk 4]
```diff
@@ -316,13 +320,11 @@ def __init__(self, options):
         self.__freshmodule = 1
         self.__curfile = None
         self.__enclosurecount = 0
+        self.__prev_token = None
 
     def __call__(self, ttype, tstring, stup, etup, line):
-        # dispatch
-##        import token
-##        print('ttype:', token.tok_name[ttype], 'tstring:', tstring,
-##              file=sys.stderr)
```
Comment (serhiy-storchaka) [inline Tools/i18n/pygettext.py]: This is not related change.

File: Lib/test/test_tools/test_i18n.py [hunk 5]
```diff
@@ -87,17 +87,26 @@ def assert_POT_equal(self, expected, actual):
...
+
+    def extract_messages_from_str(self, module_content):
+        """Return all msgids extracted from module_content."""
+        return self.extract_from_str(module_content)
```
Comment (serhiy-storchaka) [inline Lib/test/test_tools/test_i18n.py]: Why not use `extract_from_str` instead of `extract_messages_from_str`?

**Review by serhiy-storchaka — COMMENTED** [review]
Note that no warnings are emitted if option `--docstrings` is used. I think that we can use a similar approach. We can add
```
            if ttype == tokenize.NAME and tstring in ('class', 'def'):
                self.__state = self.__ignorenext
                return
```
where `__ignorenext` simply sets `self.__state = self.__waiting`.

File: Tools/i18n/pygettext.py (hunk 2 above)
Comment (tomasr8) [reply Tools/i18n/pygettext.py]: Got it! I'll revert :) Would you accept a separate (perhaps not backported) PR that removes the coding and the commented-out code or do you think it's not worth it?

File: Lib/test/test_tools/test_i18n.py (hunk 5 above)
Comment (tomasr8) [reply Lib/test/test_tools/test_i18n.py]: I wanted to mirror `extract_docstrings_from_str`, but yeah, I can directly call `extract_from_str` :+1:

File: Tools/i18n/pygettext.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Tools/i18n/pygettext.py]: I'll accept it if there are pygettext tests for files with non-UTF-8 encoding.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM. :+1:

File: Tools/i18n/pygettext.py (hunk 2 above)
Comment (tomasr8) [reply Tools/i18n/pygettext.py]: Fair enough, I'll add it to my todo list :)
